File: src/vscripts/abilities/heroes/tmp/nevermore/nevermore_ultimate/nevermore_ultimate.lua
nevermore_ultimate = class({})
LinkLuaModifier("modifier_nevermore_ultimate", "abilities/heroes/nevermore/nevermore_ultimate/modifier_nevermore_ultimate", LUA_MODIFIER_MOTION_NONE)
LinkLuaModifier("modifier_nevermore_ultimate_thinker", "abilities/heroes/nevermore/nevermore_ultimate/modifier_nevermore_ultimate_thinker", LUA_MODIFIER_MOTION_NONE)

GetAnimation(){		return GameActivity.DOTA_CAST_ABILITY_6 }
GetPlaybackRateOverride(){ 	    return 1.0 }
GetCastingCrawl(){ 			return 0 }

OnAbilityPhaseStart(){
	this.thinker = CreateModifierThinker(
		this.caster, --hCaster
		this, --hAbility
		"modifier_nevermore_ultimate_thinker", --modifierName
		{ duration = this.GetCastPoint() },
		this.caster.GetAbsOrigin(), --vOrigin
		this.caster.GetTeamNumber(), --nTeamNumber
		false --bPhantomBlocker
	)

	const particle_cast = "particles/units/heroes/hero_nevermore/nevermore_wings.vpcf"
	this.particleId = ParticleManager.CreateParticle(particle_cast, ParticleAttachment.ABSORIGIN_FOLLOW, this.caster)
	EmitGlobalSound("Hero_Nevermore.ROS.Arcana.Cast")
	return true
}

OnAbilityPhaseInterrupted(){
	this.thinker:Destroy()
	
	ParticleManager.DestroyParticle(this.particleId, true)
	ParticleManager.ReleaseParticleIndex(this.particleId)
	StopGlobalSound("Hero_Nevermore.ROS.Arcana.Cast")
}

OnSpellStart(){
	
	const origin = this.caster.GetAbsOrigin()
	const duration = this.GetSpecialValueFor("slow_duration")
	const damage = this.GetSpecialValueFor("damage_per_soul")
	const radius = this.GetSpecialValueFor("radius")
	const base_damage = this.GetSpecialValueFor("base_damage")
	const damage_per_soul = this.GetSpecialValueFor("damage_per_soul") 

	const damage_table = {
		attacker = this.caster,
		damage_type = DamageTypes.PURE,
	}

	const modifier = this.caster.FindModifierByName('modifier_nevermore_souls')
	const souls = modifier:GetStackCount()
	modifier:SetStackCount(0)

	const lines = 5 + (souls * 3)
	damage_table.damage = (base_damage + (souls * damage_per_soul))/2

	const initial_angle_deg = this.caster.GetAnglesAsVector().y
	const delta_angle = 360/lines

	for i = 0, lines - 1 do
		const facing_angle_deg = initial_angle_deg + delta_angle * i
		if(facing_angle_deg>360 ) { facing_angle_deg = facing_angle_deg - 360 }
		const facing_angle = math.rad(facing_angle_deg)
		const facing_vector = Vector(math.cos(facing_angle), math.sin(facing_angle), 0):Normalized()
		const velocity = facing_vector * 1800

        const projectile = {
            EffectName = "particles/nevermore/nevermore_basic_attack.vpcf",
            vSpawnOrigin = this.caster.GetAbsOrigin() + Vector(0, 0, 80),
            fDistance = radius,
            Source = this.caster,
            vVelocity = velocity,
            UnitBehavior = PROJECTILES_NOTHING,
            TreeBehavior = PROJECTILES_NOTHING,
            WallBehavior = PROJECTILES_NOTHING,
            GroundBehavior = PROJECTILES_NOTHING,
			fGroundOffset = 0,
            UnitTest = function(_self, unit) return unit:GetUnitName() ~= "npc_dummy_unit" and ! CustomEntitiesLegacy:Allies(_self.Source, unit) },
            OnFinish = function(_self, position)
               this.PlayEffectsOnFinish(position: Vector)
            },
        }
        ProjectilesManagerInstance:CreateProjectile(projectile)
	}

	ApplyCallbackForUnitsInArea(this.caster, origin, radius/2, DOTA_UNIT_TARGET_TEAM_ENEMY, function(enemy)
		damage_table.victim = enemy
		ApplyDamage(damage_table)
		EmitSoundOn("Hero_Spectre.Attack", enemy)
	})

	ApplyCallbackForUnitsInArea(this.caster, origin, radius, DOTA_UNIT_TARGET_TEAM_ENEMY, function(enemy)
		damage_table.victim = enemy
		ApplyDamage(damage_table)
		EmitSoundOn("Hero_Spectre.Attack", enemy)
	})

	CreateRadiusMarker(this.caster, origin, radius/2, RADIUS_SCOPE_PUBLIC, 0.1)
	CreateRadiusMarker(this.caster, origin, radius, RADIUS_SCOPE_PUBLIC, 0.1)
	ScreenShake(origin, 100, 300, 0.45, 1000, 0, true)

	ParticleManager.ReleaseParticleIndex(this.particleId)
	this.PlayEffectsLines(lines)
}

function nevermore_ultimate:PlayEffectsLines(lines)
	
	EmitSoundOn("Hero_Nevermore.ROS.Arcana", this.caster)

	const particle_cast = "particles/units/heroes/hero_nevermore/nevermore_requiemofsouls.vpcf"
	const particleId = ParticleManager.CreateParticle(particle_cast, ParticleAttachment.ABSORIGIN_FOLLOW, this.caster)
	ParticleManager.SetParticleControl(particleId, 1, Vector(lines, 0, 0))	-- Lines
	ParticleManager.SetParticleControlForward(particleId, 2, this.caster.GetForwardVector())		-- initial direction
	ParticleManager.ReleaseParticleIndex(particleId)
}

PlayEffectsOnFinish(position: Vector){
	const particle_cast = "particles/econ/items/shadow_fiend/sf_desolation/sf_base_attack_desolation_explosion.vpcf"
	const particleId = ParticleManager.CreateParticle(particle_cast, ParticleAttachment.ABSORIGIN, this.caster)
	ParticleManager.SetParticleControl(particleId, 3, position)
	ParticleManager.ReleaseParticleIndex(particleId)
}

if(IsClient() ) { require("wrappers/abilities") }
Abilities.Castpoint(nevermore_ultimate)

File: src/vscripts/abilities/heroes/nevermore/nevermore_ultimate.ts
import { registerAbility, registerModifier } from "../../../lib/dota_ts_adapter";
import { ModifierThinker, ModifierThinkerParams } from "../../../modifiers/modifier_thinker";
import { ProjectileBehavior } from "../../../projectiles";
import { direction2D } from "../../../util";
import { CustomAbility } from "../../framework/custom_ability";
import { ModifierNevermoreStacks } from "./nevermore_basic_attack";

@registerAbility("nevermore_ultimate")
class NevermoreUltimate extends CustomAbility {
    thinkerModifier?: ModifierNevermoreUltimateThinker;
    particleId?: ParticleID;

    GetAnimation() {
        return GameActivity.DOTA_CAST_ABILITY_6;
    }
    GetPlaybackRateOverride() {
        return 1.0;
    }
    GetCastingCrawl() {
        return 0;
    }

    OnAbilityPhaseStart() {
        if (super.OnAbilityPhaseStart()) {
            this.thinkerModifier = ModifierNevermoreUltimateThinker.createThinker(this.caster, this, this.caster.GetAbsOrigin(), {
                delayTime: this.GetCastPoint(),
                radius: this.GetSpecialValueFor("radius"),
                duration: this.GetSpecialValueFor("duration") + this.GetCastPoint()
            })[1];

            this.particleId = ParticleManager.CreateParticle(
                "particles/units/heroes/hero_nevermore/nevermore_wings.vpcf",
                ParticleAttachment.ABSORIGIN_FOLLOW,
                this.caster
            );
            EmitGlobalSound("Hero_Nevermore.ROS.Arcana.Cast");
            return true;
        }
        return false;
    }

    OnAbilityPhaseInterrupted() {
        super.OnAbilityPhaseInterrupted();
        this.thinkerModifier?.Destroy();

        if (this.particleId) {
            ParticleManager.DestroyParticle(this.particleId, true);
            ParticleManager.ReleaseParticleIndex(this.particleId);
        }

        StopGlobalSound("Hero_Nevermore.ROS.Arcana.Cast");
    }

    OnSpellStart() {
        this.thinkerModifier?.Destroy();

        const origin = this.caster.GetAbsOrigin();
        const radius = this.GetSpecialValueFor("radius");
        const baseDamage = this.GetSpecialValueFor("base_damage");
        const damagePerSoul = this.GetSpecialValueFor("damage_per_soul");
        const modifier = ModifierNevermoreStacks.findOne(this.caster);
        const souls = modifier?.GetStackCount() ?? 0;
        const damage = (baseDamage + souls * damagePerSoul) / 2;

        const lines = 5 + souls * 3;
        const initialAngleDeg = this.caster.GetAnglesAsVector().y;
        const deltaAngle = 360 / lines;

        for (let i = 0; i < lines; i++) {
            let facingAngleDeg = initialAngleDeg + deltaAngle * i;

            if (facingAngleDeg > 360) {
                facingAngleDeg = facingAngleDeg - 360;
            }
            const facingAngle = math.rad(facingAngleDeg);
            const facingVector = Vector(math.cos(facingAngle), math.sin(facingAngle), 0).Normalized();
            const velocity = facingVector.__mul(1800);

            this.ProjectileAttack({
                source: this.caster,
                effectName: "particles/nevermore/nevermore_basic_attack.vpcf",
                spawnOrigin: origin.__add(Vector(0, 0, 80)),
                velocity: velocity,
                groundOffset: 0,
                distance: radius,
                unitBehavior: ProjectileBehavior.NOTHING,
                wallBehavior: ProjectileBehavior.NOTHING,
                unitTest: (unit, projectile) =>
                    unit.GetUnitName() != "npc_dummy_unit" && !CustomEntitiesLegacy.Allies(projectile.getSource(), unit),
                onFinish: projectile => {
                    this.PlayEffectsOnFinish(projectile.getPosition());
                }
            });
        }

        this.DamageEnemiesInArea(origin, radius, damage);
        this.DamageEnemiesInArea(origin, radius / 2, damage);

        ScreenShake(origin, 100, 300, 0.45, 1000, 0, true);

        modifier?.SetStackCount(0);

        this.PlayEffectsLines(lines);
    }

    DamageEnemiesInArea(origin: Vector, radius: number, damage: number) {
        const enemies = CustomEntitiesLegacy.FindUnitsInRadius(
            this.caster,
            origin,
            radius,
            UnitTargetTeam.ENEMY,
            UnitTargetType.HERO + UnitTargetType.BASIC,
            UnitTargetFlags.NONE,
            FindOrder.ANY
        );

        for (const enemy of enemies) {
            ApplyDamage({
                victim: enemy,
                attacker: this.caster,
                damage: damage,
                damage_type: DamageTypes.PURE
            });
            EmitSoundOn("Hero_Spectre.Attack", enemy);
        }
    }

    PlayEffectsLines(lines: number) {
        EmitSoundOn("Hero_Nevermore.ROS.Arcana", this.caster);

        const particleId = ParticleManager.CreateParticle(
            "particles/units/heroes/hero_nevermore/nevermore_requiemofsouls.vpcf",
            ParticleAttachment.ABSORIGIN_FOLLOW,
            this.caster
        );
        ParticleManager.SetParticleControl(particleId, 1, Vector(lines, 0, 0));
        ParticleManager.SetParticleControlForward(particleId, 2, this.caster.GetForwardVector());
        ParticleManager.ReleaseParticleIndex(particleId);
    }

    PlayEffectsOnFinish(position: Vector) {
        const particleId = ParticleManager.CreateParticle(
            "particles/econ/items/shadow_fiend/sf_desolation/sf_base_attack_desolation_explosion.vpcf",
            ParticleAttachment.ABSORIGIN,
            this.caster
        );
        ParticleManager.SetParticleControl(particleId, 3, position);
        ParticleManager.ReleaseParticleIndex(particleId);
    }
}

@registerModifier("modifier_nevermore_ultimate_thinker")
class ModifierNevermoreUltimateThinker extends ModifierThinker {
    particleId?: ParticleID;

    OnCreated(params: ModifierThinkerParams) {
        super.OnCreated(params);

        if (IsServer()) {
            this.PlayEffects();
        }
    }

    OnDestroy() {
        super.OnDestroy();

        if (IsServer()) {
            this.StopEffects();
        }
    }

    PlayEffects() {
        this.particleId = ParticleManager.CreateParticle(
            "particles/units/heroes/hero_shadow_demon/shadow_demon_soul_catcher_v2_ground01.vpcf",
            ParticleAttachment.WORLDORIGIN,
            undefined
        );
        ParticleManager.SetParticleControl(this.particleId, 0, this.parent.GetAbsOrigin());
        ParticleManager.SetParticleControl(this.particleId, 1, this.parent.GetAbsOrigin());
        ParticleManager.SetParticleControl(this.particleId, 2, this.parent.GetAbsOrigin());
        ParticleManager.SetParticleControl(this.particleId, 3, Vector(this.radius, 0, 0));
    }

    StopEffects() {
        if (this.particleId) {
            ParticleManager.DestroyParticle(this.particleId, false);
            ParticleManager.ReleaseParticleIndex(this.particleId);
        }
    }

    OnIntervalThink() {
        super.OnIntervalThink();

        if (IsServer()) {
            const enemies = CustomEntitiesLegacy.FindUnitsInRadius(
                this.caster,
                this.origin,
                this.radius,
                UnitTargetTeam.ENEMY,
                UnitTargetType.HERO + UnitTargetType.BASIC,
                UnitTargetFlags.NONE,
                FindOrder.ANY
            );

            for (const enemy of enemies) {
                const direction = direction2D(enemy.GetAbsOrigin(), this.origin);
                const point = enemy.GetAbsOrigin().__add(direction.__mul(8));
                enemy.SetAbsOrigin(point);
                FindClearSpaceForUnit(enemy, enemy.GetAbsOrigin(), true);
            }
        }
    }
}
